akka-stream/src/main/scala/akka/stream/impl/SetupStage.scala

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl
+
+import akka.annotation.InternalApi
+import akka.stream.scaladsl.{ Flow, Keep, Sink, Source }
+import akka.stream._
+import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, InHandler, OutHandler }
+
+import scala.concurrent.{ Future, Promise }
+
+/** Internal Api */
+@InternalApi private[stream] final class SetupSinkStage[T, M](factory: ActorMaterializer ⇒ Attributes ⇒ Sink[T, M])
+  extends GraphStageWithMaterializedValue[SinkShape[T], Future[M]] {
+
+  private val in = Inlet[T]("SetupSinkStage.in")
+  override val shape = SinkShape(in)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[M]) = {
+    val matPromise = Promise[M]
+    (createStageLogic(matPromise), matPromise.future)
+  }
+
+  private def createStageLogic(matPromise: Promise[M]) = new GraphStageLogic(shape) {
+    import SetupStage._
+
+    val subOutlet = new SubSourceOutlet[T]("SetupSinkStage")
+    subOutlet.setHandler(delegateToInlet(() ⇒ pull(in), () ⇒ cancel(in)))
+    setHandler(in, delegateToSubOutlet(() ⇒ grab(in), subOutlet))
+
+    override def preStart(): Unit = {
+      val sink = factory(ActorMaterializerHelper.downcast(materializer))(attributes)
+
+      val mat = Source.fromGraph(subOutlet.source).runWith(sink.withAttributes(attributes))(subFusingMaterializer)
+      matPromise.success(mat)
+    }
+  }
+
+}
+
+/** Internal Api */
+@InternalApi private[stream] final class SetupFlowStage[T, U, M](factory: ActorMaterializer ⇒ Attributes ⇒ Flow[T, U, M])
+  extends GraphStageWithMaterializedValue[FlowShape[T, U], Future[M]] {
+
+  private val in = Inlet[T]("SetupFlowStage.in")
+  private val out = Outlet[U]("SetupFlowStage.out")
+  override val shape = FlowShape(in, out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[M]) = {
+    val matPromise = Promise[M]
+    (createStageLogic(matPromise), matPromise.future)
+  }
+
+  private def createStageLogic(matPromise: Promise[M]) = new GraphStageLogic(shape) {
+    import SetupStage._
+
+    val subInlet = new SubSinkInlet[U]("SetupFlowStage")
+    val subOutlet = new SubSourceOutlet[T]("SetupFlowStage")
+
+    subInlet.setHandler(delegateToOutlet(push(out, _: U), () ⇒ complete(out), fail(out, _), subInlet))
+    subOutlet.setHandler(delegateToInlet(() ⇒ pull(in), () ⇒ cancel(in)))
+
+    setHandler(in, delegateToSubOutlet(() ⇒ grab(in), subOutlet))
+    setHandler(out, delegateToSubInlet(subInlet))
+
+    override def preStart(): Unit = {
+      val flow = factory(ActorMaterializerHelper.downcast(materializer))(attributes)
+
+      val mat = Source
+        .fromGraph(subOutlet.source)
+        .viaMat(flow.withAttributes(attributes))(Keep.right)
+        .to(Sink.fromGraph(subInlet.sink))
+        .run()(subFusingMaterializer)
+      matPromise.success(mat)
```

Comment:
`Try` and complete the promise with that in case materialization factory or materialization fails?

Reply:
Good point. I will also add test-cases for those situations.